openvdb/cmd/openvdb_print.cc

```diff
@@ -11,13 +11,116 @@
+
+    for (auto leafIter = inputGrid->tree().cbeginLeaf(); leafIter; ++leafIter) {
+        auto attrset = leafIter->attributeSet();
+        auto dptr = attrset.descriptorPtr();
+        auto attrmap = dptr->map();
+
+        points.total += leafIter->pointCount();
+        points.active += leafIter->onPointCount();
+        points.inactive += leafIter->offPointCount();
+
+        // groups
+        auto grmap = dptr->groupMap();
+
+        // Count points in groups
+        for (auto it=grmap.begin(); it!=grmap.end(); ++it){
+            if(points.groups.find(it->first) == points.groups.end()){
+                points.groups[it->first] = 0;
+            }
+            points.groups[it->first]+=leafIter->groupPointCount(it->first);
+        }
+
+        for (auto it = grmap.begin(); it!=grmap.end(); ++it){
+            std::cout << it->first << std::endl;
+        }
+        // Collect attribute info
+        for (auto it=attrmap.begin(); it!=attrmap.end(); ++it){
+            PointAttrib pa{};
+            pa.name = it->first;
+            pa.index = it->second;
+            pa.type = dptr->valueType(pa.index);
+            pa.isShared = attrset.isShared(pa.index);
+
+            auto attrArr = attrset.getConst(pa.name);
+            pa.codec = attrArr->codecType();
+            pa.isUniform = attrArr->isUniform();
+            pa.isHidden = attrArr->isHidden();
+            pa.isTransient = attrArr->isTransient();
+            pa.isStreaming = attrArr->isStreaming();
+            points.pointAttribs[pa.name] = pa;
+        }
+    }
+}
+
+static void
+printPointStats(const Points& pointStats)
+{
+        std::cout << "Total Point Count:\n"
+            << sINDENT << "total: " << pointStats.total << '\n'
+            << sINDENT << "active: " << pointStats.active  << '\n'
+            << sINDENT << "inactive: " << pointStats.inactive << '\n';
+        std::cout << "Point attributes:" << '\n';
+        for(auto it=pointStats.pointAttribs.begin(); it!=pointStats.pointAttribs.end(); ++it){
+            auto attr = it->second;
+            std::cout << "name: " << attr.name << '\n';
+            std::cout << sINDENT << "index: " << attr.index << '\n';
+            std::cout << sINDENT << "type: " << attr.type << '\n';
+            std::cout << sINDENT << "codec: " << attr.codec << '\n';
+            std::cout << sINDENT << "isUniform: " << attr.isUniform << '\n';
+            std::cout << sINDENT << "isShared: " << attr.isShared << '\n';
+            std::cout << sINDENT << "isHidden: " << attr.isHidden << '\n';
+            std::cout << sINDENT << "isTransient: " << attr.isTransient << '\n';
+            std::cout << sINDENT << "isStreaming: " << attr.isStreaming << '\n';
+          }
+        std::cout << "Point groups:" << '\n';
+        for (auto it=pointStats.groups.begin(); it!=pointStats.groups.end(); ++it){
+            std::cout << it->first << " " << it->second <<std::endl;
+        }
+}
```

Comment:
Can you provide an example print output which we can comment on w.r.t formatting

Reply:
```
...
Transform:
  voxel size: 2.29
  index to world:
     [2.29, 0, 0, 0] 
     [0, 2.29, 0, 0] 
     [0, 0, 2.29, 0] 
     [0, 0, 0, 1] 

Total Point Count:
   total: 8
   active: 8
   inactive: 0
Point attributes:
  name: P
   type: vec3s
   codec: fxpt16
   uniform: 1
   shared: 1
   hidden: 0
   transient: 0
   group: 0
   string: 0
   constantStride: 1
  name: __group
   type: vec3s
   codec: grp
   uniform: 0
   shared: 1
   hidden: 0
   transient: 0
   group: 0
   string: 0
   constantStride: 1
Point groups:
   group1: 5
   group2: 6
```

Reply:
I suggest we do a first pass of this where we don't include any details that require iterating over the AttributeArrays, but the ideal in my mind would be something like the following:

name: P
   type: vec3s
   codec: fxpt16
   stride: 1 (or stride: dynamic)
   hidden: 0/123 *
   transient: 5/123 *
   uniform: 100/123 *

The group and string fields should be evident from the value type / codec, so I'd drop those fields entirely. Also, I'd make "type" -> "valueType" just to eliminate confusion there.

The fields marked with asterisks I would probably only display if in verbose mode (but as I say let's just leave that out for now). These show the number of arrays that are uniform (ie 100) vs the total number of arrays across all leaf nodes (ie 123).

You'll note I've changed "constantStride" -> "stride". I think it's better to mark it as "dynamic" if it's not constant. Otherwise it's potentially confusing, does constantStride = 1 mean _true_ it's constant stride or that the stride is 1.

Reply:
I just realized changes I've pushed in the meantime haven't made into this PR as the original PR was opened from a develop not a feature branch in my fork. Sorry about that.
@danrbailey I will add changes as suggested in your last comment. 
One thing which is not clear is to me is 
> we don't include any details that require iterating over the AttributeArrays,
I believe this refers to the new Info::Array

Reply:
Well stride is available in Info::Array or by accessing the AttributeArray directly, I'm more talking about those fields such as uniform which are _not_ available in Info::Array. For now, I suggest we just ignore the starred ones in my list above.

Reply:
Ok, I've addressed all comments, I hope. Below is the the current output:

```
...
Total Point Count:
   total: 8
   active: 8
   inactive: 0
Point attributes:
  name: P
   valueType: vec3s
   codec: fxpt16
   stride: 1
  name: __group
   valueType: vec3s
   codec: grp
   stride: 1
Point groups:
   group1: 5
   group2: 6
```

To double check, we still interested only to inspect the first leaf for attributes? (assuming there is consistency across all other leaves)